--warn-unreachable false positive with minimal example without types explicitly set

Hello,

I just downloaded the new mypy v0.720 and am playing with the `--warn-unreachable-flag` in my projects.

I noticed one occasion where there is something that seems like a false positive, so I am making an issue for it with a minimal reproducible example.


#### Version
```
mypy==0.720
mypy-extensions==0.4.1
```

#### source code example

```python
import sys
from typing import Optional

found_id = None

symbols_list = [{'symbol': 'A', 'id': 1}, {'symbol': 'B', 'id': 2},  {'symbol': 'B', 'id': 3}]

for entry in symbols_list:
    if entry['symbol'] == 'B':
        if found_id:
            print('error, symbol found multiple times!')
            sys.exit(1)

        found_id = entry['id']
```


This when tested with `mypy test.py --warn-unreachable` returns:
> test.py:10: error: Statement is unreachable

So essentially telling us that `if found_id` is never entered due to the `found_id` type being `None`. But in reality the type should be `Optional[int]` since you can see that it's set to `entry['id']` later.

If you explicitly set the types mypy no longer complains:
```python
import sys
from typing import Optional

found_id: Optional[int] = None

symbols_list = [{'symbol': 'A', 'id': 1}, {'symbol': 'B', 'id': 2},  {'symbol': 'B', 'id': 3}]

for entry in symbols_list:
    if entry['symbol'] == 'B':
        if found_id:
            print('error, symbol found multiple times!')
            sys.exit(1)

        assert isinstance(entry['id'], int)
        found_id = entry['id']
```

#### Bug or not

Essentially I am not sure if this is intended behaviour or not. After writing this issue down I realize it all boils down to the matter of mypy not correctly inferring type of `found_id`.

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -584,14 +584,23 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         *,
         exit_condition: Expression | None = None,
     ) -> None:
-        """Repeatedly type check a loop body until the frame doesn't change.
-        If exit_condition is set, assume it must be False on exit from the loop.
+        """Repeatedly type check a loop body until the frame doesn't change."""
 
-        Then check the else_body.
-        """
-        # The outer frame accumulates the results of all iterations
+        # The outer frame accumulates the results of all iterations:
         with self.binder.frame_context(can_skip=False, conditional_frame=True):
+
+            # Check for potential decreases in the number of partial types so as not to stop the
+            # iteration too early:
             partials_old = sum(len(pts.map) for pts in self.partial_types)
+
+            # Disable error types that we cannot safely identify in intermediate iteration steps:
+            warn_unreachable = self.options.warn_unreachable
+            if warn_unreachable:
+                self.options.warn_unreachable = False
+            warn_redundant = codes.REDUNDANT_EXPR in self.options.enabled_error_codes
+            if warn_redundant:
+                self.options.enabled_error_codes.remove(codes.REDUNDANT_EXPR)
+
             while True:
                 with self.binder.frame_context(can_skip=True, break_frame=2, continue_frame=1):
                     self.accept(body)
@@ -599,9 +608,20 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 if (partials_new == partials_old) and not self.binder.last_pop_changed:
                     break
                 partials_old = partials_new
+
+            # If necessary, reset the modified options and make up for the postponed error checks:
+            if warn_unreachable or warn_redundant:
+                self.options.warn_unreachable = warn_unreachable
+                self.options.enabled_error_codes.add(codes.REDUNDANT_EXPR)
+                with self.binder.frame_context(can_skip=True, break_frame=2, continue_frame=1):
+                    self.accept(body)
+
+            # If exit_condition is set, assume it must be False on exit from the loop:
             if exit_condition:
                 _, else_map = self.find_isinstance_check(exit_condition)
                 self.push_type_map(else_map)
+
+            # Check the else body:
             if else_body:
                 self.accept(else_body)
 

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -595,11 +595,9 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
             # Disable error types that we cannot safely identify in intermediate iteration steps:
             warn_unreachable = self.options.warn_unreachable
-            if warn_unreachable:
-                self.options.warn_unreachable = False
             warn_redundant = codes.REDUNDANT_EXPR in self.options.enabled_error_codes
-            if warn_redundant:
-                self.options.enabled_error_codes.remove(codes.REDUNDANT_EXPR)
+            self.options.warn_unreachable = False
+            self.options.enabled_error_codes.discard(codes.REDUNDANT_EXPR)
 
             while True:
                 with self.binder.frame_context(can_skip=True, break_frame=2, continue_frame=1):
@@ -610,10 +608,10 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 partials_old = partials_new
 
             # If necessary, reset the modified options and make up for the postponed error checks:
+            self.options.warn_unreachable = warn_unreachable
+            if warn_redundant:
+                self.options.enabled_error_codes.add(codes.REDUNDANT_EXPR)
             if warn_unreachable or warn_redundant:
-                self.options.warn_unreachable = warn_unreachable
-                if warn_redundant:
-                    self.options.enabled_error_codes.add(codes.REDUNDANT_EXPR)
                 with self.binder.frame_context(can_skip=True, break_frame=2, continue_frame=1):
                     self.accept(body)
 

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -612,7 +612,8 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
             # If necessary, reset the modified options and make up for the postponed error checks:
             if warn_unreachable or warn_redundant:
                 self.options.warn_unreachable = warn_unreachable
-                self.options.enabled_error_codes.add(codes.REDUNDANT_EXPR)
+                if warn_redundant:
+                    self.options.enabled_error_codes.add(codes.REDUNDANT_EXPR)
                 with self.binder.frame_context(can_skip=True, break_frame=2, continue_frame=1):
                     self.accept(body)
 

--- a/test-data/unit/check-narrowing.test
+++ b/test-data/unit/check-narrowing.test
@@ -2390,3 +2390,29 @@ class A:
                 z.append(1)
 
 [builtins fixtures/primitives.pyi]
+
+[case testAvoidFalseUnreachableInLoop]
+# flags: --warn-unreachable --python-version 3.11
+
+def f() -> int | None: ...
+def b() -> bool: ...
+
+x: int | None
+x = 1
+while x is not None or b():
+    x = f()
+
+[builtins fixtures/bool.pyi]
+
+[case testAvoidFalseRedundantExprInLoop]
+# flags: --enable-error-code redundant-expr --python-version 3.11
+
+def f() -> int | None: ...
+def b() -> bool: ...
+
+x: int | None
+x = 1
+while x is not None and b():
+    x = f()
+
+[builtins fixtures/primitives.pyi]

